### PR TITLE
support markdown message type and @someone

### PR DESCRIPTION
**Pull request by happywzy**
1) support markdown message when use dingtalk chatbot. `kubectl edit nm notification-manager`, config the receivers:
```yaml
dingtalk:
  msgType: markdown
```
2) support @someone when use dingtalk chatbot. message (text or markdown) content need contain phone number like `@131xxxxx899`

### CODE REVIEW

**Comment by happywzy** [comment]
image
image
image

**Review by wanjunlei — CHANGES_REQUESTED** [review]

File: pkg/apis/v2beta2/notificationmanager_types.go [hunk 1]
```diff
@@ -188,6 +188,8 @@ type DingTalkOptions struct {
 	// The name of the template to generate DingTalk message.
 	// If the global template is not set, it will use default.
 	Template string `json:"template,omitempty"`
+	// message type: text or markdown
+	MsgType string `json:"msgType,omitempty"`
```
[the remark this answers is not in the document]
Comment (benjaminhuo) [reply pkg/apis/v2beta2/notificationmanager_types.go:192]: MsgType or TmplType, it looks more like a template type?
@wanjunlei you mean only the dingtalk can use markdown template?

**Comment by wanjunlei** [comment]
Thanks for your commit, can you make the dingtalk conversation support markdown type too?

**Comment by happywzy** [comment]
> Thanks for your commit, can you make the dingtalk conversation support markdown type too?

The dingtalk conversation message format : [official link](https://developers.dingtalk.com/document/app/send-group-messages)
```json
{
        "chatid":"chate39f5xxxxxx335",
        "msg":{
                "msgtype":"text",
                "text":{
                        "content":"请于本月底提交月度工作报告。"
                }
        },        
}
```
chatbot message format: [link](https://developers.dingtalk.com/document/app/custom-robot-access)
```json
{
    "at": {
        "atMobiles":[
            "180xxxxxx"
        ],
        "atUserIds":[
            "user123"
        ],
        "isAtAll": false
    },
    "text": {
        "content":"我就是我, @XXX 是不一样的烟火"
    },
    "msgtype":"text"
}
```
but in the func `sendToChatBot()` and  `sendToConversation()`, the message format  has no special treatment, I think this may be a bug? I need test first.

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

**Comment by benjaminhuo** [comment]
@happywzy I've discussed with @wanjunlei , we agree that you can keep your original changes to the global DingTalk options.

And you can add the same option to the local DingTalkReceiver as a per receiver option. https://github.com/kubesphere/notification-manager/blob/master/pkg/apis/v2beta2/receiver_types.go#L40

- If the global DingTalk TmplType is not set and the local DingTalk TmplType is set, the local DingTalk TmplType will be used.
- If the global DingTalk TmplType is set and the local DingTalk TmplType is set too, the local DingTalk TmplType will be used.
- If the global DingTalk TmplType is set and the local DingTalk TmplType is not set, the global DingTalk TmplType will be used.

The MsgType should be changed to TmplType

**Comment by happywzy** [comment]
> @happywzy I've discussed with @wanjunlei , we agree that you can keep your original changes to the global DingTalk options.
> 
> And you can add the same option to the local DingTalkReceiver as a per receiver option. https://github.com/kubesphere/notification-manager/blob/master/pkg/apis/v2beta2/receiver_types.go#L40
> 
> * If the global DingTalk TmplType is not set and the local DingTalk TmplType is set, the local DingTalk TmplType will be used.
> * If the global DingTalk TmplType is set and the local DingTalk TmplType is set too, the local DingTalk TmplType will be used.
> * If the global DingTalk TmplType is set and the local DingTalk TmplType is not set, the global DingTalk TmplType will be used.
> 
> The MsgType should be changed to TmplType

OK, Please look at this question 
https://github.com/kubesphere/notification-manager/pull/84#issuecomment-871026698

**Comment by benjaminhuo** [comment]
> > Thanks for your commit, can you make the dingtalk conversation support markdown type too?
> 
> The dingtalk conversation message format : [official link](https://developers.dingtalk.com/document/app/send-group-messages)
> 
> ```json
> {
>         "chatid":"chate39f5xxxxxx335",
>         "msg":{
>                 "msgtype":"text",
>                 "text":{
>                         "content":"请于本月底提交月度工作报告。"
>                 }
>         },        
> }
> ```
> 
> chatbot message format: [link](https://developers.dingtalk.com/document/app/custom-robot-access)
> 
> ```json
> {
>     "at": {
>         "atMobiles":[
>             "180xxxxxx"
>         ],
>         "atUserIds":[
>             "user123"
>         ],
>         "isAtAll": false
>     },
>     "text": {
>         "content":"我就是我, @XXX 是不一样的烟火"
>     },
>     "msgtype":"text"
> }
> ```
> 
> but in the func `sendToChatBot()` and `sendToConversation()`, the message format has no special treatment, I think this may be a bug? I need test first.

@wanjunlei is this a bug ?

**Comment by wanjunlei** [comment]
> > Thanks for your commit, can you make the dingtalk conversation support markdown type too?
> 
> The dingtalk conversation message format : [official link](https://developers.dingtalk.com/document/app/send-group-messages)
> 
> ```json
> {
>         "chatid":"chate39f5xxxxxx335",
>         "msg":{
>                 "msgtype":"text",
>                 "text":{
>                         "content":"请于本月底提交月度工作报告。"
>                 }
>         },        
> }
> ```
> 
> chatbot message format: [link](https://developers.dingtalk.com/document/app/custom-robot-access)
> 
> ```json
> {
>     "at": {
>         "atMobiles":[
>             "180xxxxxx"
>         ],
>         "atUserIds":[
>             "user123"
>         ],
>         "isAtAll": false
>     },
>     "text": {
>         "content":"我就是我, @XXX 是不一样的烟火"
>     },
>     "msgtype":"text"
> }
> ```
> 
> but in the func `sendToChatBot()` and `sendToConversation()`, the message format has no special treatment, I think this may be a bug? I need test first.


I did not understand what you mean, can you explain it in more detail？

**Comment by happywzy** [comment]
> > > Thanks for your commit, can you make the dingtalk conversation support markdown type too?
> > 
> > 
> > The dingtalk conversation message format : [official link](https://developers.dingtalk.com/document/app/send-group-messages)
> > ```json
> > {
> >         "chatid":"chate39f5xxxxxx335",
> >         "msg":{
> >                 "msgtype":"text",
> >                 "text":{
> >                         "content":"请于本月底提交月度工作报告。"
> >                 }
> >         },        
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > chatbot message format: [link](https://developers.dingtalk.com/document/app/custom-robot-access)
> > ```json
> > {
> >     "at": {
> >         "atMobiles":[
> >             "180xxxxxx"
> >         ],
> >         "atUserIds":[
> >             "user123"
> >         ],
> >         "isAtAll": false
> >     },
> >     "text": {
> >         "content":"我就是我, @XXX 是不一样的烟火"
> >     },
> >     "msgtype":"text"
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but in the func `sendToChatBot()` and `sendToConversation()`, the message format has no special treatment, I think this may be a bug? I need test first.
> 
> I did not understand what you mean, can you explain it in more detail？

https://github.com/kubesphere/notification-manager/blob/ec47ba1f5eb83cca3fe99b03f263c2654cd94f38/pkg/notify/notifier/dingtalk/dingtalk.go#L355
```java
dm := dingtalkMessage{
	Text: dingtalkMessageContent{
		Content: msg,
	},
	Type: "text",
	ID:   chatID,
}
```
this data structure has problem, if you want send message by Conversation.

File: pkg/notify/notifier/dingtalk/dingtalk.go [hunk 2]
```diff
@@ -119,6 +139,11 @@ func NewDingTalkNotifier(logger log.Logger, receivers []config.Receiver, notifie
 			n.templateName = opts.Global.Template
 		}
 
+		if len(d.TmplType) > 0 && strings.Compare("markdown", d.TmplType) == 0 {
```
Comment (wanjunlei) [inline pkg/notify/notifier/dingtalk/dingtalk.go:142]: You can refer to [this](https://github.com/kubesphere/notification-manager/blob/master/pkg/notify/notifier/dingtalk/dingtalk.go#L116) to implement the following logic.

- If the global DingTalk TmplType is not set and the local DingTalk TmplType is set, the local DingTalk TmplType will be used.
- If the global DingTalk TmplType is set and the local DingTalk TmplType is set too, the local DingTalk TmplType will be used.
- If the global DingTalk TmplType is set and the local DingTalk TmplType is not set, the global DingTalk TmplType will be used.

File: pkg/notify/notifier/dingtalk/dingtalk.go [hunk 3]
```diff
@@ -215,25 +240,53 @@ func (n *Notifier) sendToChatBot(ctx context.Context, d *config.DingTalk, data t
...
-			Text: dingtalkMessageContent{
-				Content: msg,
-			},
-		}
-
 		var buf bytes.Buffer
-		if err := utils.JsonEncode(&buf, dm); err != nil {
-			_ = level.Error(n.logger).Log("msg", "DingTalkNotifier: encode message error", "error", err.Error())
-			return err
+		if strings.Compare("markdown", d.TmplType) == 0 || (len(d.TmplType) == 0 && strings.Compare("markdown", n.tmplType) == 0) {
```
Comment (happywzy) [inline pkg/notify/notifier/dingtalk/dingtalk.go:259]: here is the logic  you need @wanjunlei

File: pkg/notify/notifier/dingtalk/dingtalk.go (hunk 2 above)
Comment (happywzy) [reply pkg/notify/notifier/dingtalk/dingtalk.go:142]: this is logic to set global DingTalk TmplType  from config, default is text.

File: pkg/notify/notifier/dingtalk/dingtalk.go (hunk 3 above)
Comment (wanjunlei) [reply pkg/notify/notifier/dingtalk/dingtalk.go:259]: Ok

**Review by wanjunlei — APPROVED** [review]

**Review by benjaminhuo — APPROVED** [review]

**Comment by benjaminhuo** [comment]
@happywzy We would like to invite you as the maintainer of Notification Manager.

**Comment by FeynmanZhou** [comment]
Do we need to add the document for these two features in the [Official Documentation](https://kubesphere.io/zh/docs/cluster-administration/platform-settings/notification-management/configure-dingtalk/)?

**Comment by benjaminhuo** [comment]
Yeah，should add to docs after Notification Manager release a new version or ks v3.2 is released

**Pull request merged**